### PR TITLE
qemu_guest_agent: new case of test guest-fstrim for windows vm

### PR DESCRIPTION
**Pull request by xiagao**
ID: 1566802
Signed-off-by: Xiaoling Gao <xiagao@redhat.com>

### CODE REVIEW

**Comment by xiagao** [comment]
(1/1) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.qemu_guest_agent.virtio_serial.check_fstrim.q35: PASS (93.34 s)

**Comment by xiagao** [comment]
@xiangchunfu  Call for your review.. Thanks.

File: qemu/tests/qemu_guest_agent.py [hunk 1]
```diff
@@ -2043,6 +2020,95 @@ def result_check(flag, write_timeout, session):
...
+        data_file = os.path.join(guest_dir,
+                                 "qga_fstrim%s" %
+                                 utils_misc.generate_random_string(5))
+        for i in range(5):
+            count = 1000 * (i + 1)
+            logging.info("Create %sM file in guest." % count)
+            cmd = "dd if=/dev/zero of=%s bs=1M count=%d" % (data_file, count)
+            session.cmd(cmd, timeout=600)
+            delete_file_cmd = "%s %s" % (params["delete_file_cmd"],
+                                         data_file.replace("/", "\\"))
```
Comment (xiangchunfu) [inline qemu/tests/qemu_guest_agent.py]: @xiagao  For windows,  This command( "del /f /q". ) is completely remove file from disk?  If it only put file to the recycling center.   Then it don't reduce disk size.

File: qemu/tests/qemu_guest_agent.py [hunk 2]
```diff
@@ -2043,6 +2020,95 @@ def result_check(flag, write_timeout, session):
...
+
+        error_context.context("Create fragment in data disk.", logging.info)
+        guest_dir = r"%s:" % mnt_point[0]
+        data_file = os.path.join(guest_dir,
+                                 "qga_fstrim%s" %
+                                 utils_misc.generate_random_string(5))
+        for i in range(5):
+            count = 1000 * (i + 1)
+            logging.info("Create %sM file in guest." % count)
+            cmd = "dd if=/dev/zero of=%s bs=1M count=%d" % (data_file, count)
```
Comment (xiangchunfu) [inline qemu/tests/qemu_guest_agent.py]: @xiagao,  Does windows guest support dd to create file?

File: qemu/tests/qemu_guest_agent.py [hunk 3]
```diff
@@ -2043,6 +2020,95 @@ def result_check(flag, write_timeout, session):
...
+        :param params: Dictionary with the test parameters
+        :param env: Dictionary with test environment.
+        """
+        def get_blocks():
+            """
+            Get the used blocks of data disk.
+            :return: the used blocks
+            """
+            blocks = process.system_output("stat -t %s" % stg_image_filename)
+            return blocks.strip().split()[2]
```
Comment (xiangchunfu) [inline qemu/tests/qemu_guest_agent.py]: @xiagao,  This command can get real data size for block device or file image?

File: qemu/tests/qemu_guest_agent.py (hunk 1 above)
Comment (xiagao) [reply qemu/tests/qemu_guest_agent.py]: I think it's removed from disk.And the following code verify it.

File: qemu/tests/qemu_guest_agent.py (hunk 2 above)
Comment (xiagao) [reply qemu/tests/qemu_guest_agent.py]: yes.

File: qemu/tests/qemu_guest_agent.py (hunk 3 above)
Comment (xiagao) [reply qemu/tests/qemu_guest_agent.py]: file image according to polarion steps.

File: qemu/tests/qemu_guest_agent.py [hunk 4]
```diff
@@ -2043,6 +2020,95 @@ def result_check(flag, write_timeout, session):
...
+
+        error_context.context("Execute the guest-fstrim cmd via qga.",
+                              logging.info)
+        self.gagent.fstrim()
+
+        error_context.context("Check blocks of data disk after fstrim.",
+                              logging.info)
+        blocks_after_fstrim = get_blocks()
+
+        if int(blocks_after_fstrim) > int(blocks_before_fstrim):
```
Comment (xiangchunfu) [inline qemu/tests/qemu_guest_agent.py]: ＠xiagao  should >=,   when (blocks_after_fstrim) = int(blocks_before_fstrim), fstrim doesn't works. It should be fail.

File: qemu/tests/qemu_guest_agent.py [hunk 5]
```diff
@@ -2043,6 +2020,95 @@ def result_check(flag, write_timeout, session):
...
+            session.cmd(cmd, timeout=600)
+            delete_file_cmd = "%s %s" % (params["delete_file_cmd"],
+                                         data_file.replace("/", "\\"))
+            logging.info("Delete the guest file created just now.")
+            session.cmd(delete_file_cmd)
+
+        error_context.context("Check blocks of data disk before fstrim.",
+                              logging.info)
+        blocks_before_fstrim = get_blocks()
+        if int(blocks_init) > int(blocks_before_fstrim):
```
Comment (xiangchunfu) [inline qemu/tests/qemu_guest_agent.py]: ＠xiagao  should >=

File: qemu/tests/qemu_guest_agent.py (hunk 2 above)
Comment (xiangchunfu) [reply qemu/tests/qemu_guest_agent.py]: ＠xiagao,  If possible,  should use /dev/random,  shouldn't use /dev/zero.   /dev/zero is another format disk.

Comment (xiagao) [reply qemu/tests/qemu_guest_agent.py]: You are right, update.

File: qemu/tests/qemu_guest_agent.py (hunk 4 above)
Comment (xiagao) [reply qemu/tests/qemu_guest_agent.py]: Modified.

File: qemu/tests/qemu_guest_agent.py (hunk 5 above)
Comment (xiagao) [reply qemu/tests/qemu_guest_agent.py]: Update

**Comment by xiangchunfu** [comment]
Acked-by:Xiangchun Fu xfu@redhat.com

**Comment by xiagao** [comment]
@yanan-fu  Call for your review, thanks.

File: qemu/tests/cfg/qemu_guest_agent.cfg [hunk 6]
```diff
@@ -219,9 +231,6 @@
             cmd_get_disk = cat /proc/mounts |awk '$2~/^\%s$/{print $1,$3}'
             Windows:
                 cmd_get_disk = wmic volume where "DriveLetter='%s'" get FileSystem,DeviceID |findstr /v /i FileSystem
-        - check_nonexistent_cmd:
```
Comment (yanan-fu) [inline qemu/tests/cfg/qemu_guest_agent.cfg]: This is a new cases that be added few days ago, could you explain why did you delete it?

Comment (yanan-fu) [reply qemu/tests/cfg/qemu_guest_agent.cfg]: https://github.com/autotest/tp-qemu/pull/1876

Comment (xiagao) [reply qemu/tests/cfg/qemu_guest_agent.cfg]: @yanan-fu I'm not sure the reason, maybe misoperation or rebase issue.
Anyway ,add this config part again.
thanks.

**Comment by xiagao** [comment]
@yanan-fu Call for review when you are available,thanks.

File: qemu/tests/qemu_guest_agent.py [hunk 7]
```diff
@@ -2140,6 +2142,95 @@ def result_check(flag, write_timeout, session):
...
+            :return: the used blocks
+            """
+            blocks = process.system_output("stat -t %s" % stg_image_filename)
+            return blocks.strip().split()[2]
+
+        session = self._get_session(params, None)
+        self._open_session_list.append(session)
+
+        error_context.context("Format data disk.", logging.info)
+        image_size_disk = params["image_size_stg"]
```
Comment (yanan-fu) [inline qemu/tests/qemu_guest_agent.py]: Prefer `image_size_stg`

File: qemu/tests/qemu_guest_agent.py [hunk 8]
```diff
@@ -2140,6 +2142,95 @@ def result_check(flag, write_timeout, session):
...
+        disk_index = utils_misc.wait_for(
```
Comment (yanan-fu) [inline qemu/tests/qemu_guest_agent.py]: `disk_index_stg` ?

File: qemu/tests/qemu_guest_agent.py [hunk 9]
```diff
@@ -2140,6 +2142,95 @@ def result_check(flag, write_timeout, session):
...
+        if not utils_disk.update_windows_disk_attributes(session, disk_index):
+            test.error("Failed to update windows disk attributes.")
+        mnt_point = utils_disk.configure_empty_disk(
+            session, disk_index[0], image_size_disk, "windows",
+            labeltype="msdos")
+
+        error_context.context("Check the original blocks of data disk.",
+                              logging.info)
+        stg_image_params = params.object_params("stg")
+        stg_image_filename = storage.get_image_filename(
```
Comment (yanan-fu) [inline qemu/tests/qemu_guest_agent.py]: For those params: `image_params`, `image_filename`, `image_size`,`disk_index`. 
`**_stg` or `stg_**` are all ok for me, but suggest to keep consistent.

File: qemu/tests/qemu_guest_agent.py [hunk 10]
```diff
@@ -2140,6 +2142,95 @@ def result_check(flag, write_timeout, session):
...
+        Execute "guest-fstrim" command via guest agent.
+
+        Steps:
+        1) boot up guest with scsi backend device
+        2) init and format the data disk
+        3) get the original blocks of data disk
+        4) create fragment in data disk
+        5) check the used blocks
+        6) execute fstrim cmd via guest agent
+        7) check if the used blocks is decreased
```
Comment (yanan-fu) [inline qemu/tests/qemu_guest_agent.py]: I am not sure the necessity of checking different disk format, as polarion test both qcow2 and raw.

File: qemu/tests/qemu_guest_agent.py (hunk 7 above)
Comment (xiagao) [reply qemu/tests/qemu_guest_agent.py]: update

File: qemu/tests/qemu_guest_agent.py (hunk 8 above)
Comment (xiagao) [reply qemu/tests/qemu_guest_agent.py]: disk_index is ok

File: qemu/tests/qemu_guest_agent.py (hunk 9 above)
Comment (xiagao) [reply qemu/tests/qemu_guest_agent.py]: their name is unified. Thanks.

**Comment by xiagao** [comment]
@yanan-fu  Update,pls have a check. Thanks.
(1/1) Host_RHEL.m8.u1.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.qemu_guest_agent.virtio_serial.check_fstrim: PASS (144.44 s)

**Comment by yanan-fu** [comment]
LGTM

Acked-by: Yanan Fu <yfu@redhat.com>
